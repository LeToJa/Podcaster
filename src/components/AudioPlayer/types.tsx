export interface AudioPlayerTypes {
    url: string
    type: string
}