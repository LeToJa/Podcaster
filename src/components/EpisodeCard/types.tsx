export interface EpisodeCardTypes {
    podcastId: string
    episodeId: number
    title: string
    date: string
    duration: string
}