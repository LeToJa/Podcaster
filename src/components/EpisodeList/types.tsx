import { PodcastEpisode } from '../../layout/PodcastLayout/types'

export interface EpisodeListTypes {
    podcastId: string
    episodes: PodcastEpisode[]
}