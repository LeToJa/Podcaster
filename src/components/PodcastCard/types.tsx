export interface PocastCardTypes {
    id: string
    name: string
}