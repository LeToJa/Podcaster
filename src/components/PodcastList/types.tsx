import { PocastCardTypes } from '../PodcastCard/types'

export interface PodcastListTypes {
    podcasts: PocastCardTypes[]
}