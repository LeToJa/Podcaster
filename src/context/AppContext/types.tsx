export interface AppContextTypes {
    loading: boolean
    toggleLoading: (value: boolean) => void
}