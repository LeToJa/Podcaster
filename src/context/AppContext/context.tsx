import { createContext, useState } from 'react'

import { ReactElementWithChildrenTypes } from '../../types'
import { AppContextTypes } from './types'

export const AppContext = createContext<AppContextTypes | undefined>(undefined)

export const AppProvider = ({ children }: ReactElementWithChildrenTypes) => {
    const [loading, setLoading] = useState<boolean>(true)

    const toggleLoading = (value: boolean) => setLoading(value)

    return <>
        <AppContext.Provider value={{ loading, toggleLoading }}>
            {children}
        </AppContext.Provider>
    </>
}