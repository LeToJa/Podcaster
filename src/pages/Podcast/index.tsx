import { useEffect, useState } from 'react'
import { useParams } from 'react-router-dom'

import { useApp } from '../../context/AppContext'

import { getPodcastInfo } from '../../helpers/ItunesAPI'

import PodcastLayout from '../../layout/PodcastLayout'
import EpisodeList from '../../components/EpisodeList'

import { PodcastAsideTypes, PodcastEpisode } from '../../layout/PodcastLayout/types'

const Podcast = () => {
    const { podcastId } = useParams()
    const { loading, toggleLoading } = useApp()
    const [aside, setAside] = useState<PodcastAsideTypes | false>(false)
    const [episodes, setEpisodes] = useState<PodcastEpisode[]>()

    useEffect(() => {
        const fetchPodcasts = async () => {
            const content = await getPodcastInfo(String(podcastId))

            if (!content) {
                setAside(false)
                toggleLoading(false)
                return
            }

            setAside({
                id: content.id,
                title: content.title,
                author: content.author,
                artwork: content.artwork,
                description: content.description,
            })

            setEpisodes(content.episodes)

            toggleLoading(false)
        }

        toggleLoading(true)
        fetchPodcasts().catch(err => console.error(err))
    }, [podcastId])

    if (loading) return

    if (aside === false) return <section className='bg-gray-50 border-2 border-gray-200 shadow-xl p-4' data-testid='error-message'>This podcast doesn&apos;t work :(</section>

    return <>
        <PodcastLayout aside={aside}>
            <EpisodeList podcastId={podcastId!} episodes={episodes!} />
        </PodcastLayout>
    </>
}

export default Podcast