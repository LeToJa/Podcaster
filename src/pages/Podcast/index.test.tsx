import { render, waitFor } from '@testing-library/react'
import { describe, it, expect, vi, beforeEach } from 'vitest'
import { BrowserRouter, Params } from 'react-router-dom'

import { AppContext } from '../../context/AppContext/context'
import { getPodcastInfo } from '../../helpers/ItunesAPI'

import { AppContextProps } from '../../context/AppContext/types'
import { GenericObject, ReactElementWithChildrenTypes } from '../../types'
import { PodcastItemsTypes } from '../../layout/PodcastLayout/types'

import Podcast from '.'

vi.mock('react-router-dom', async (importOriginal): Promise<unknown> => {
    const actual: GenericObject = await importOriginal()

    return {
        ...actual,
        useParams: (): Readonly<Params<string>> => ({ podcastId: '123' }),
    }
})

vi.mock('../../helpers/ItunesAPI', () => ({
    getPodcastInfo: vi.fn(),
}))

const mockAppContext = {
    loading: false,
    toggleLoading: vi.fn()
}

const MockAppProvider = ({ loading, toggleLoading, children }: AppContextProps & ReactElementWithChildrenTypes) => {
    return (
        <AppContext.Provider value={{ loading, toggleLoading }}>
            {children}
        </AppContext.Provider>
    )
}

const mockPodcast: PodcastItemsTypes | false = {
    id: '123',
    title: 'Sample Podcast',
    author: 'Author Name',
    artwork: 'some-url.jpg',
    description: 'Podcast Description',
    episodes: [
        {
            title: 'Episode 1',
            content: '<p>Episode content in HTML format</p>',
            date: '2023-08-29T12:00:00Z',
            url: 'http://example.com/audio.mp3',
            type: 'audio/mpeg',
            duration: '30:00'
        }
    ]
}

const mockGetPodcastInfo = vi.mocked(getPodcastInfo)

describe('Podcast', () => {
    beforeEach(() => {
        mockGetPodcastInfo.mockResolvedValue(mockPodcast)
    })

    it('renders correctly and matches snapshot', async () => {
        const { container, getByText } = render(
            <BrowserRouter>
                <MockAppProvider {...mockAppContext}>
                    <Podcast />
                </MockAppProvider>
            </BrowserRouter>
        )

        await waitFor(() => {
            expect(getByText(/Sample Podcast/i)).toBeTruthy()
        })

        expect(container).toMatchSnapshot()
    })

    it('renders "This Podcast doesnt work" when there is no podcast data', async () => {
        mockGetPodcastInfo.mockResolvedValueOnce(undefined)

        const { getByText } = render(
            <BrowserRouter>
                <MockAppProvider {...mockAppContext}>
                    <Podcast />
                </MockAppProvider>
            </BrowserRouter>
        )

        await waitFor(() => {
            expect(getByText(/This Podcast doesn't work/i)).toBeTruthy()
        })
    })

    it('renders podcast layout and episode list when there is podcast data', async () => {
        const { getByText } = render(
            <BrowserRouter>
                <MockAppProvider {...mockAppContext}>
                    <Podcast />
                </MockAppProvider>
            </BrowserRouter>
        )

        await waitFor(() => {
            expect(getByText(/Sample Podcast/i)).toBeTruthy()
            expect(getByText(/Episode 1/i)).toBeTruthy()
        })
    })
})
