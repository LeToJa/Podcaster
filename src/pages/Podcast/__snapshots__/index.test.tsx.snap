// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Podcast > renders correctly and matches snapshot 1`] = `
<div>
  <section
    class="grid grid-cols-7 gap-12"
  >
    <aside
      class="col-span-2"
    >
      <figure
        class="bg-gray-50 border-2 border-gray-200 shadow-xl p-4"
        data-testid="podcast-aside"
      >
        <img
          alt="Sample Podcast"
          class="w-4/5 mx-auto rounded-full"
          src="some-url.jpg"
        />
        <header
          class="p-4"
        >
          <a
            href="/podcast/123"
          >
            <h4
              class="font-bold text-xl mb-2"
            >
              Sample Podcast
            </h4>
            <h6
              class="text-gray-700 text-base"
            >
              Author Name
            </h6>
          </a>
        </header>
        <blockquote
          class="text-sm"
        >
          <h5
            class="font-bold text-lg mb-2"
          >
            Description
          </h5>
          Podcast Description
        </blockquote>
      </figure>
    </aside>
    <article
      class="col-span-5"
      data-testid="podcast-main"
    >
      <figure
        class="bg-gray-50 border-2 border-gray-200 shadow-xl p-4 mb-6"
        data-testid="podcast-count"
      >
        <h4
          class="text-xl"
        >
          1
           episodes
        </h4>
      </figure>
      <figure
        class="bg-gray-50 border-2 border-gray-200 shadow-xl p-4"
        data-testid="podcast-episodes"
      >
        <table
          class="w-full table-fixed"
        >
          <thead>
            <tr>
              <th
                colspan="8"
              >
                Title
              </th>
              <th>
                Date
              </th>
              <th>
                Duration
              </th>
            </tr>
          </thead>
          <tbody>
            <tr
              data-testid="podcast-episode"
            >
              <td
                colspan="8"
              >
                <a
                  class="font-bold text-sky-600"
                  href="/podcast/123/episode/0"
                >
                  Episode 1
                </a>
              </td>
              <td
                class="text-center"
              >
                2023-08-29T12:00:00Z
              </td>
              <td
                class="text-center"
              >
                30:00
              </td>
            </tr>
          </tbody>
        </table>
      </figure>
    </article>
  </section>
</div>
`;
